File: 3. Behavioral Design Patterns/Strategy/Kotlin/StrategyTemplate.kt
// Strategy Pattern – Kotlin Template

interface Strategy { fun execute(data: List<Int>): List<Int> }
class ConcreteStrategyA : Strategy { override fun execute(data: List<Int>) = data }
class ConcreteStrategyB : Strategy { override fun execute(data: List<Int>) = data }

class Context(private var strategy: Strategy) {
    fun setStrategy(s: Strategy) { strategy = s }
    fun doWork(data: List<Int>) = strategy.execute(data)
}

fun main(){ val ctx = Context(ConcreteStrategyA()); ctx.doWork(emptyList()) }

File: 3. Behavioral Design Patterns/Strategy/C#/StrategyTemplate.cs
// Strategy Pattern – C# Template

using System.Collections.Generic;

namespace StrategyTemplate
{
    interface IStrategy { IList<int> Execute(IList<int> data); }
    class ConcreteStrategyA : IStrategy { public IList<int> Execute(IList<int> data){ return data; } }
    class ConcreteStrategyB : IStrategy { public IList<int> Execute(IList<int> data){ return data; } }

    class Context
    {
        private IStrategy _strategy;
        public Context(IStrategy strategy){ _strategy = strategy; }
        public void SetStrategy(IStrategy strategy){ _strategy = strategy; }
        public IList<int> DoWork(IList<int> data) => _strategy.Execute(data);
    }

    class Program { static void Main(){ var ctx = new Context(new ConcreteStrategyA()); ctx.DoWork(new List<int>()); } }
}
